CRM/Mailing/BAO/Mailing.php

```diff
@@ -226,34 +226,43 @@ public static function getRecipients($mailingID) {
     );
 
     if ($isSMSmode) {
-      $includeFilters = array(
-        "mg.group_type = 'Include'",
-        'mg.search_id IS NULL',
-        "$contact.is_opt_out = 0",
-        "$contact.is_deceased <> 1",
-        "$entityTable.phone_type_id = " . CRM_Core_PseudoConstant::getKey('CRM_Core_DAO_Phone', 'phone_type_id', 'Mobile'),
-        "$entityTable.phone IS NOT NULL",
-        "$entityTable.phone != ''",
-        "$entityTable.is_primary = 1",
-        "mg.mailing_id = #mailingID",
-        'temp.contact_id IS null',
+      $params = array(
+        'filters' => array(
+          'is_opt_out' => "$contact.is_opt_out = 0",
+          'is_deceased' => "$contact.is_deceased <> 1",
+          'location_filter' => "$entityTable.phone_type_id = " . CRM_Core_PseudoConstant::getKey('CRM_Core_DAO_Phone', 'phone_type_id', 'Mobile'),
+          'phone_not_null' => "$entityTable.phone IS NOT NULL",
+          'phone_not_empty' => "$entityTable.phone != ''",
+          'is_primary' => "$entityTable.is_primary = 1",
+          'mailing_id' => "mg.mailing_id = #mailingID",
+          'temp_contact_null' => 'temp.contact_id IS null',
+        ),
+        'order_by' => array(
+          "$entityTable.is_primary = 1",
+        ),
       );
-      $order_by = array("$entityTable.is_primary = 1");
     }
     else {
       // Criterias to filter recipients that need to be included
-      $includeFilters = array(
-        "$contact.do_not_email = 0",
-        "$contact.is_opt_out = 0",
-        "$contact.is_deceased <> 1",
-        $location_filter,
-        "$entityTable.email IS NOT NULL",
-        "$entityTable.email != ''",
-        "mg.mailing_id = #mailingID",
-        'temp.contact_id IS NULL',
+      $params = array(
+        'filters' => array(
+          'do_not_email' => "$contact.do_not_email = 0",
+          'is_opt_out' => "$contact.is_opt_out = 0",
+          'is_deceased' => "$contact.is_deceased <> 1",
+          'location_filter' => $location_filter,
+          'email_not_null' => "$entityTable.email IS NOT NULL",
+          'email_not_empty' => "$entityTable.email != ''",
+          'mailing_id' => "mg.mailing_id = #mailingID",
+          'temp_contact_null' => 'temp.contact_id IS NULL',
+        ),
+        'order_by' => $order_by,
       );
     }
 
+    // Allow user to alter query responsible to fetch mailing recipients before build,
+    //   by changing the mail filters identified $params
+    CRM_Utils_Hook::alterMailingRecipients($mailingObj, $params, 'pre');
```

Comment:
I have mixed feelings here -- mostly over what makes a good contract for this. First consideration: *do SQL expressions represent a supportable/maintainable contract?*

On one hand, you might say "no" because there's many possible symbols in SQL. I like how `$params['filters']` gives a *name* to each condition; the list of names is short and concrete. If you want to remove a criterion, you can do that without understanding the particular SQL involved.

On the other hand, this hook doesn't really hide the SQL -- if you want to add a new condition or change a condition, then you'll have to understand the list of tables/aliases/fields in play. And if you're going to have that responsibility, then we should probably go all the way: pass a `CRM_Utils_SQL_Select` object to give access to the full list of params/joins/groupings/subqueries.

As an alternative, consider changing the list of SQL strings (`$params['filters']`) to a list of SQL objects -- i.e. https://github.com/totten/civicrm-core/commit/6995c3e4587a34dd3da4903cb393c96ee84cdd70 (with https://github.com/totten/civicrm-core/commit/3ed9641888fb3415e7ff59bca87a27245572d6da). This allows the custom filter to benefit from escaping data, joining, etc.

```php
<?php
// Ex: Remove a criterion by name
unset($criteria['is_deceased']);

// Ex: Add a filter based on subtype. Use some data provided by the site admin (with string-escaping).
$criteria['my_extra_criterion'] = CRM_Utils_SQL_Select::fragment()
  ->where('civicrm_contact.contact_sub_type IN (@subtypes)')
  ->param('subtypes', Civi::settings()->get('myextension_mailable_subtypes'));

// Ex: Add a filter based on Drupal role. Use some data provided by the admin (with numeric-validation).
$criteria['my_extra_criterion'] = CRM_Utils_SQL_Select::fragment()
  ->join('civicrm_uf_match', 'INNER JOIN civicrm_uf_match WHERE civicrm_uf_match.contact_id = civicrm_contact.id')
  ->join('users_roles', 'INNER JOIN users_roles ON civicrm_uf_match.uf_id = users_roles.uid AND rid IN (#allowedRoles)')
  ->param('allowedRoles', Civi::settings()->get('myextension_drupal_role_ids'));
```

Of course, you need to know/assume some things about SQL query. (Ex: You assume that the `civicrm_contact` record is named `civicrm_contact` and not `contact_a` or `contact_b` or `recipient`.) But that was also true when passing SQL strings.